src/MongoDB.Driver.Core/Core/Misc/ServerVersion.cs

```diff
@@ -0,0 +1,463 @@
+            : this(major, minor, patch, null, internalBuild, commitHash)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ServerVersion"/> class.
+        /// </summary>
+        /// <param name="major">The major version.</param>
+        /// <param name="minor">The minor version.</param>
+        /// <param name="patch">The patch version.</param>
+        /// <param name="releaseCandidate">The release candidate version.</param>
+        /// <param name="internalBuild">The internal build version.</param>
+        /// <param name="commitHash">The internal build commit hash.</param>
+        public ServerVersion(int major, int minor, int patch, int? releaseCandidate, int? internalBuild, string commitHash)
+        {
+            _major = Ensure.IsGreaterThanOrEqualToZero(major, nameof(major));
+            _minor = Ensure.IsGreaterThanOrEqualToZero(minor, nameof(minor));
+            _patch = Ensure.IsGreaterThanOrEqualToZero(patch, nameof(patch));
+            _releaseCandidate = releaseCandidate; // can be null
+            _internalBuild = internalBuild; // can be null
+            _commitHash = commitHash; // can be null
+        }
+
+        // properties
+        /// <summary>
+        /// Gets the internal build commit hash.
+        /// </summary>
+        /// <value>
+        /// The internal build commit hash.
+        /// </value>
+        public string CommitHash
+        {
+            get { return _commitHash; }
+        }
+
+        /// <summary>
+        /// Gets the internal build version.
+        /// </summary>
+        /// <value>
+        /// The internal build version.
+        /// </value>
+        public int? InternalBuild
+        {
+            get { return _internalBuild; }
+        }
+
+        /// <summary>
+        /// Gets the major version.
+        /// </summary>
+        /// <value>
+        /// The major version.
+        /// </value>
+        public int Major
+        {
+            get { return _major; }
+        }
+
+        /// <summary>
+        /// Gets the minor version.
+        /// </summary>
+        /// <value>
+        /// The minor version.
+        /// </value>
+        public int Minor
+        {
+            get { return _minor; }
+        }
+
+        /// <summary>
+        /// Gets the patch version.
+        /// </summary>
+        /// <value>
+        /// The patch version.
+        /// </value>
+        public int Patch
+        {
+            get { return _patch; }
+        }
+
+        /// <summary>
+        /// Gets the release candidate version.
+        /// </summary>
+        /// <value>
+        /// The release candidate version.
+        /// </value>
+        public int? ReleaseCandidate
+        {
+            get { return _releaseCandidate; }
+        }
+
+        // public methods
+        public int CompareTo(ServerVersion other)
+        {
+            if (ReferenceEquals(other, null))
```

Comment:
do we really need ReferenceEquals for asserting on null?

Reply:
Yes. Otherwise this would trigger `==` operator overload.

Reply:
Nits:

If it it worked in SemanticVersion we don't necessarily need to change this. If it is important enough to change then SemanticVersion should in theory be changed also.

We usually use `object.ReferenceEquals` instead of just `ReferenceEquals` for clarity.

Reply:
In `SemanticVersion` this triggered an unnecessary overloads in several places.
I can improve the logic of `SemanticVersion` if necessary.
Moreover `SemanticVersion` has a incorrect `null` comparison:
```
SemanticVersion a = null;
SemanticVersion b = null;
var t = a > b; // true
```